torch/hub.py

```diff
@@ -120,28 +121,57 @@ def _read_url(url):
 
 
 def _validate_not_a_forked_repo(repo_owner, repo_name, branch):
+    # Here we try to make sure that the branch/commit hash effectively belongs
+    # to the repo_owner/repo_name repo.
+    # This is important because GitHub is weird, and the download URL
+    # f'https://github.com/{repo_owner}/{repo_name}/archive/{commit_hash}.zip'
+    # may actually exist *even if* commit_hash is from a different repo_owner / fork.
+    # For example this URL exists:
+    # https://github.com/pytorch/vision/archive/8949c7011facf7801fdf077cc3e4ecd8f0940c7e.zip
+    # even though the commit hash doesn't come from pytorch/vision. It actually comes from a fork
+    # https://github.com/NicolasHug/vision/commit/8949c7011facf7801fdf077cc3e4ecd8f0940c7e
+    # An interesting example of this is https://news.ycombinator.com/item?id=24882921.
+    # So we want avoid downloading code that can come from a potentionally malicious fork.
+
     # Use urlopen to avoid depending on local git.
     headers = {'Accept': 'application/vnd.github.v3+json'}
     token = os.environ.get(ENV_GITHUB_TOKEN)
     if token is not None:
         headers['Authorization'] = f'token {token}'
-    for url_prefix in (
-            f'https://api.github.com/repos/{repo_owner}/{repo_name}/branches',
-            f'https://api.github.com/repos/{repo_owner}/{repo_name}/tags'):
-        page = 0
-        while True:
-            page += 1
-            url = f'{url_prefix}?per_page=100&page={page}'
-            response = json.loads(_read_url(Request(url, headers=headers)))
-            # Empty response means no more data to process
-            if not response:
-                break
-            for br in response:
-                if br['name'] == branch or br['commit']['sha'].startswith(branch):
-                    return
+    try:
+        for url_prefix in (
```

Comment:
For reviewers: I didn't change this block at all apart from the error message at the end. I just put it inside a try/except statement so everything is indented to the right

Reply:
I think this is a neat update and it surely improves the user experience. However the cyclomatic complexity of this whole `try~except~raise` is very high. So I suggest to extract the logic around `while True` block, as a follow up.

Reply:
thanks, that makes sense. I extracted the "find branch" logic into a separate subfunction, hopefully the code looks simpler now